edge-modules/MetricsCollector/Program.cs

```diff
@@ -0,0 +1,54 @@
+// Copyright (c) Microsoft. All rights reserved.
+namespace MetricsCollector
+{
+    using System;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Microsoft.Azure.Devices.Client;
+    using Microsoft.Azure.Devices.Client.Transport.Mqtt;
+    using Microsoft.Azure.Devices.Edge.Agent.Diagnostics;
+    using Microsoft.Azure.Devices.Edge.Agent.Diagnostics.Publisher;
+    using Microsoft.Azure.Devices.Edge.ModuleUtil;
+    using Microsoft.Azure.Devices.Edge.Util;
+    using Microsoft.Extensions.Logging;
+
+    internal class Program
+    {
+        static readonly ILogger Logger = ModuleUtil.CreateLogger("MetricsCollector");
+
+        public static int Main() => MainAsync().Result;
+
+        private static async Task<int> MainAsync()
+        {
+            Logger.LogInformation($"Starting metrics collector with the following settings:\r\n{Settings.Current}");
+
+            MetricsScraper scraper = new MetricsScraper(Settings.Current.Endpoints);
+            IMetricsPublisher publisher;
+            if (Settings.Current.UploadTarget == UploadTarget.AzureLogAnalytics)
+            {
+                publisher = new LogAnalyticsUpload(Settings.Current.AzMonWorkspaceId, Settings.Current.AzMonWorkspaceKey, Settings.Current.AzMonLogType, Guid.NewGuid());
+            }
+            else
+            {
+                MqttTransportSettings mqttSetting = new MqttTransportSettings(TransportType.Mqtt_Tcp_Only);
+                ITransportSettings[] transportSettings = { mqttSetting };
+                ModuleClient moduleClient = await ModuleClient.CreateFromEnvironmentAsync(transportSettings);
+                publisher = new EventHubMetricsUpload(moduleClient);
+            }
+
+            (CancellationTokenSource cts, ManualResetEventSlim completed, Option<object> handler) = ShutdownHandler.Init(TimeSpan.FromSeconds(5), Logger);

```

Comment:
why this line is called between metrics publisher and metrics scrap & upload?  should it put before metrics publisher?

Reply:
I wanted to keep it as close as possible to where it is needed. It is not needed before this point because we don't use the cts or other objects until the body of the using statement and after.